### ISSUE TITLE
unproject image to pointcloud in pytorch3d

### ISSUE THREAD

**Issue by jjjkkyz** [facebookresearch/pytorch3d#1709]
I am working to generate pointcloud from image with pytorch3d.
As mentioned in https://pytorch3d.org/docs/cameras , I make a demo:
```
image_size = 768
fcl_screen = image_size
prp_screen = ((fcl_screen,fcl_screen),)
R, T = look_at_view_transform(dist, elev, azim)
image_size = torch.tensor([image_size, image_size]).unsqueeze(0)
cameras = PerspectiveCameras(focal_length=fcl_screen, principal_point=prp_screen, R=R, T=T, device=device, image_size=image_size, in_ndc=False)

x = torch.linspace(image_size-1,0, image_size)
y = torch.linspace(image_size-1,0, image_size )
y, x = torch.meshgrid(x, y)
x = x.unsqueeze(-1).cuda()
y = y.unsqueeze(-1).cuda()

xy_depth = torch.cat((x,y , fragments.zbuf[0]), dim=-1).reshape((-1,3))
xyz = cameras[i].unproject_points(xy_depth,world_coordinates=True)
```
Depth image is generated by rendering an input mesh.

My question is: should I use ```torch.linspace(image_size-1,0, image_size)``` or ```torch.linspace(image_size,1, image_size)``` to match the coordinate with fragments.zbuf?
And why ```torch.linspace(image_size-1,0, image_size)``` work well but ```torch.linspace(0,image_size-1, image_size)``` get a wrong result? It seems that fragment is not formed as the same?

**Comment by bottler** [comment]
The function `get_rgbd_point_cloud` from `pytorch3d.implictron.tools.point_cloud_utils` (see [here](https://github.com/facebookresearch/pytorch3d/blob/main/pytorch3d/implicitron/tools/point_cloud_utils.py)) is the recommended built-in way to do unprojection. It gets the detail right. 

(I think pytorch3d cameras always measure screen space from the outside of outer pixels, ie align_corners=True, so some -1s are needed.)

**Comment by jjjkkyz** [comment]
> The function `get_rgbd_point_cloud` from `pytorch3d.implictron.tools.point_cloud_utils` (see [here](https://github.com/facebookresearch/pytorch3d/blob/main/pytorch3d/implicitron/tools/point_cloud_utils.py)) is the recommended built-in way to do unprojection. It gets the detail right.
> 
> (I think pytorch3d cameras always measure screen space from the outside of outer pixels, ie align_corners=True, so some -1s are needed.)

Thanks for reply. By quick test code in ```get_rebd_point_cloud```, I found that:
If I project the unproject pointcloud back to camera, I get coordinate that is image_height-0.5, image_height-1.5 ... 1.5, 0.5. So, is that means pytorch3d render a image by sample in X.5 coordinate in screen space(```torch.linspace(image_size-0.5,0.5, image_size)```)? Not as I post in above which is ```torch.linspace(image_size-1,0, image_size)```

**Comment by jjjkkyz** [comment]
Meanwhile, project pointcloud back to depth image by ```cameras.transform_points(pts_3d)``` does not match the input depth image, what may cause this?

**Comment by jjjkkyz** [comment]
As I want to get a function(and its inverse function) bettween rendered image and it's unproject point cloud, simply use build-in function may not be enough. I rewrite my question below and thanks for any reply.

1. When pytorch3d render a pixel, what is its coordinate in camera space? (ie [0,1,2...image_size-1], [image_size, image_size-1...1,0], [0.5,1.5...image_size-0.5]).
2. Is ```camera.transform_points``` and ```camera.unproject_points``` meet my requirements to be a function(and its inverse function) bettween rendered image and it's unproject point cloud? Or I need some other function such as ```ray_bundle_to_ray_points``` used in ```get_rgbd_point_cloud```.

**Comment by bottler** [comment]
Re 2, for a given camera type it should be possible with unproject_points. One problem with unproject_points is that its behaviour is not completely consistent between camera types. 

Re 1, the edge of a pixel is the edge of the image, so in simple cases the pixel centers are at 0.5, 1.5 etc.

**Comment by jjjkkyz** [comment]
Sure, I sample a grid of point in camera A, unproject then to word and then transform to camera B. It seems that points do not lie in correct place in camera B. How can I solve it.

**Comment by jjjkkyz** [comment]
In details, I have a list of cameras, which is surrounding a mesh. pixels can be unproject and transform correct in therir own cameras. But resulting in a wrong coordinate in others cameras. 
For example, sample a pixels in camera A's center(50, 50, 0.5), unproject it to world and transform to camera B,C,D... (```cameras[1].transform_points(cameras[0].unproject_points(torch.FloatTensor([[50,50,0.5]]).cuda()))```), results always have a negtive value in position. AS all cameras is surrounding a mesh, any camera's center should be projected in other camera's bounding box (0~image_size ), it should not happen.

**Comment by jjjkkyz** [comment]
I found it is because I have ndc and screen camera in the same list. So, I should use ```torch.linspace(image_size-0.5,0.5, image_size)/image_size*2-1`` for ndc camera?

**Comment by bottler** [comment]
> I found it is because I have ndc and screen camera in the same list. So, I should use ```torch.linspace(image_size-0.5,0.5, image_size)/image_size*2-1`` for ndc camera?

(Easiest not to mix ndc and non-ndc.) Assuming a square image, the ndc of the (centre of the) edge pixels is probably `1-(1/image_size)` and `-1+(1/image_size)`, which matches that formula yes.

**Comment by guochengqian** [comment]
hi @bottler @jjjkkyz  I am working on the same problem, given RGBD, camera1 of image 1, and would like to transform to image 2. I met the same problem that can not generate rendering align to image 2.  I am using the NeRF lego testing set with ground truth depth as example. The code can faithfully reconstruct image 1 from the generated point cloud, but could only generate all black image in image 2.  The issue is with some coordinates convention, however I have checked all code and cannot find which part made it wrong. 

The significant part of the code is: 
```

def render_point_cloud_pytorch3d(
    camera,
    point_cloud,
    render_size: Tuple[int, int],
    **kwargs,
):
    raster_settings = PointsRasterizationSettings(
        image_size=render_size, 
        radius = 1.0 / render_size[0] * 2.0, 
        points_per_pixel = 10
    )
    rasterizer = PointsRasterizer(cameras=camera, raster_settings=raster_settings)
    renderer = PointsRenderer(
        rasterizer=rasterizer,
        compositor=AlphaCompositor()
    )

    images = renderer(point_cloud)
    rgb = images[0, :, :, :3]
    return (rgb.cpu().numpy() * 255).astype(np.uint8)

os.makedirs('debug', exist_ok=True)
nerf_data_dir = '../../data/nerf_synthetic/' 
dataset = 'lego'

# resize image
device = 'cpu'
image_size = (64, 64)
_, _, test_dataset = get_nerf_datasets(dataset, image_size)
print(len(test_dataset))

# query two image, input and reference (ground truth)
idx1, idx2 = 10, 25
frame1, frame2 = test_dataset[idx1], test_dataset[idx2]

# load depth data (pytorch3d data does not contain depth, here, load depth from original nerf synthetic dataset)
depth_path = os.path.join(nerf_data_dir, dataset, f'test/r_{idx1}_depth_0001.png')
depth = read_depth2tensor(depth_path, resize=image_size).to(device)

image = frame1['image'].to(device)
focal_length = frame1['camera'].focal_length 
camera = PerspectiveCameras(device=device, 
                            R=frame1['camera'].R, T=frame1['camera'].T,
                            focal_length=focal_length)

# unproject using function from: pytorch3d.implicitron.tools.point_cloud_utils.get_rgbd_point_cloud 
imh, imw = depth.shape
# convert the depth maps to point clouds using the grid ray sampler
pts_3d = ray_bundle_to_ray_points(
    NDCMultinomialRaysampler(
        image_width=imw,
        image_height=imh,
        n_pts_per_ray=1,
        min_depth=1.0,
        max_depth=1.0,
        unit_directions=False,
    )(camera)._replace(lengths=depth[None, ..., None])
)
pts_3d = pts_3d.reshape(-1, 3)
point_cloud = Pointclouds(points=pts_3d[None], features=image.reshape(1, -1, 3))

# # render the input
rendering_image1 = render_point_cloud_pytorch3d(camera, point_cloud, image_size)
Image.fromarray((rendering_image1).astype(np.uint8)).save("debug/rendering_image1.jpg")

# render the renference
camera_new = PerspectiveCameras(device=device, 
                            R=frame2['camera'].R, T=frame2['camera'].T,
                            focal_length=focal_length) 
rendering_image2 = render_point_cloud_pytorch3d(camera_new, point_cloud, image_size)
Image.fromarray((rendering_image2).astype(np.uint8)).save("debug/rendering2_image2.jpg")

```
![out](https://github.com/facebookresearch/pytorch3d/assets/48788073/b9720ef9-bbd4-4fa1-ac93-ab3acd7b7eb8)

**Comment by guochengqian** [comment]
Hi all, my issue was fixed in #1712. Thanks.
